Fix Travis build check

jrfnl opened the pull request:
This PR covers the following:
* Update `dev-lib` dependency.
* Test against more PHP versions (linting).
* Add a properly configured PHPCS ruleset specific to this plugin and run PHPCS only once per build.
    The ruleset is largely based on the ruleset which comes with `dev-lib`.
    Differences:
    - (temporarily) Exclude the nonce verification sniff. As this plugin works on the plugins page and WP AFAIK does nonce verification for the same query variables as being used by this plugin, a duplicate nonce check is probably not needed, but this needs further investigation to be sure.
    - Add the WPCS recommended configuration properties to optimize the results.
* Fix coding standards violations based on WPCS `develop`.
* Minor miscellaneous code simplifications.

The PR will be easiest to review by looking at the individual commits.

The build currently does not (yet) pass. The reason for this is that the `PHPCompatibility` ruleset cannot be found.

As the PHPCompatibility ruleset is also included in the example `phpcs.xml` file included in `dev-lib`, there must be a way to pull it in, but I need some help to get this working. /cc @westonruter 

Lastly, as I haven't been able to see the Travis results yet, I'm not sure whether warnings will fail the build or not. If so, the `Generic.Files.OneClassPerFile` sniff should be excluded as well, or alternatively, the plugin file should be split up in individual class files.

Fixes #50

jrfnl committed:
Update dev-lib dependency

jrfnl committed:
Build/PHPCS: add custom ruleset

jrfnl committed:
CS: comply with up-to-date coding standards

jrfnl committed:
PHPCS: whitelist usage of WP core hooks

jrfnl committed:
CS: various documentation fixes

* Fix punctuation of comments
* Fix layout of comments and docblocks

This commit does *not* address missing documentation.

jrfnl committed:
CS: minor code simplifications

jrfnl committed:
CS: change global function to class method

The global `html()` function was as it is for historic reasons.

As the function is only used by one class anyway, it may as well be moved into that class and out of the global namespace.

jrfnl committed:
CS: rename private method

Using an underscore to prefix a private method is a PHP 4 left-over and no longer needed.

As it is a private method, renaming this method is safe.

jrfnl committed:
Build/PHPCS: temporarily exclude an error which needs further investigation

jrfnl committed:
Build/Travis: update Travis script & environment variables

jrfnl committed:
Readme: lead by example - use correct block comment format

westonruter commented on .dev-lib:
```diff
@@ -0,0 +1,8 @@
+DEFAULT_BASE_BRANCH=develop
+PHPCS_GIT_TREE=master
+PHPCS_IGNORE='tests/*,dev-lib/*,includes/vendor/*'
+PHPCS_RULESET_FILE=./.phpcs.xml.dist
```
Why not just rename `.phpcs.xml.dist` to `phpcs.xml.dist` so that PHPCS will pick it up by default?

jrfnl replied:
Because PHPCS (as of v 3.1.0) will pick up `.phpcs.xml.dist` files by default as well.... With the `.` prefix, the file now groups itself with other similar configuration files, so when given the choice, using the `.` prefix is recommended.

As a side-note: the order of preference by PHPCS (as of 3.1.1) is `.phpcs.xml`, `phpcs.xml`, `.phpcs.xml.dist`, `phpcs.xml.dist`.
So using the `.` also means the file will be found sooner.

westonruter replied:
Yeah, I missed your other PR: https://github.com/xwp/wp-dev-lib/pull/266

jrfnl replied:
No problem. Glad that's cleared up ;-)

westonruter committed:
Install PHPCS, WPCS, and PHPCompatibility via Composer

westonruter commented:
@jrfnl ok, Travis is correctly running PHPCS now, so the build is now just failing due to actual coding errors/warnings that are flagged by PHPCS. I'll leave whether or not they should be fixed up to you, or just excluded as you noted.

```
plugin-dependencies.php:66:1: warning - Best practice suggestion: Declare only one class in a file. (Generic.Files.OneClassPerFile.MultipleFound)
plugin-dependencies.php:207:23: error - Processing form data without nonce verification. (WordPress.CSRF.NonceVerification.NoNonceVerification)
plugin-dependencies.php:207:69: error - Processing form data without nonce verification. (WordPress.CSRF.NonceVerification.NoNonceVerification)
plugin-dependencies.php:207:100: error - Processing form data without nonce verification. (WordPress.CSRF.NonceVerification.NoNonceVerification)
plugin-dependencies.php:207:133: error - Processing form data without nonce verification. (WordPress.CSRF.NonceVerification.NoNonceVerification)
plugin-dependencies.php:208:21: error - Processing form data without nonce verification. (WordPress.CSRF.NonceVerification.NoNonceVerification)
plugin-dependencies.php:420:23: error - Processing form data without nonce verification. (WordPress.CSRF.NonceVerification.NoNonceVerification)
plugin-dependencies.php:420:72: error - Processing form data without nonce verification. (WordPress.CSRF.NonceVerification.NoNonceVerification)
plugin-dependencies.php:420:104: error - Processing form data without nonce verification. (WordPress.CSRF.NonceVerification.NoNonceVerification)
plugin-dependencies.php:420:137: error - Processing form data without nonce verification. (WordPress.CSRF.NonceVerification.NoNonceVerification)
plugin-dependencies.php:421:21: error - Processing form data without nonce verification. (WordPress.CSRF.NonceVerification.NoNonceVerification)
plugin-dependencies.php:561:1: warning - Best practice suggestion: Declare only one class in a file. (Generic.Files.OneClassPerFile.MultipleFound)
plugin-dependencies.php:676:43: warning - json_encode() is discouraged. Use wp_json_encode() instead. (WordPress.WP.AlternativeFunctions.json_encode_json_encode)
```

westonruter commented on .dev-lib:
```diff
@@ -0,0 +1,8 @@
+DEFAULT_BASE_BRANCH=develop
+PHPCS_GIT_TREE=master
+PHPCS_IGNORE='vendor/*,tests/*,dev-lib/*,includes/vendor/*'
+PHPCS_RULESET_FILE=./.phpcs.xml.dist
```
GitHub dropped my PR comment. I think it would be better to name this just `phpcs.xml.dist` so that `phpcs` can be run from the directory without having to pass the `--standard`.

jrfnl replied:
Actually, GH didn't drop your comment, but published it as a stand-alone comment. See my response above.

jrfnl commented on .dev-lib:
```diff
@@ -1,8 +1,8 @@
 DEFAULT_BASE_BRANCH=develop
 PHPCS_GIT_TREE=master
```
As Composer is being used now, I would presume that we can drop the `PHPCS_GIT_TREE` and `WPCS_GIT_TREE` environment variables  in the `.dev-lib` file ?

jrfnl commented on composer.json:
```diff
@@ -0,0 +1,13 @@
+    "license": "GPL-2.0",
+    "version": "1.3.0",
+    "require-dev": {
+        "wp-coding-standards/wpcs": "^0.14.0",
```
While I fully support only using stable releases, in this case, `develop` would be the branch we need, see my other comment.

westonruter replied:
OK, I just copied this from another `composer.json`.

jrfnl reviewed (commented):
> the build is now just failing due to actual coding errors/warnings that are flagged by PHPCS. I'll leave whether or not they should be fixed up to you, or just excluded as you noted.

In the original setup, I set the `WPCS_GIT_TREE` to the `develop` branch because of a bug/missing feature in WPCS.
The `wp_json_encode()` function which is mentioned in the `AlternativeFunctions` warning is not available in the minimum WP version this plugin supports, so this can't be fixed.
Based on seeing the warning here, I send in a PR to WPCS to address this - https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards/pull/1285 (merged yesterday), which is why I set the WPCS branch to be used to `develop`.

As in the next version of WPCS `1.0.0`, some sniff names will change as per https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards/pull/1242, the exclude for the `NonceVerification` is not being picked up when using WPCS 0.14.0.

That just leaves the `OneClassPerFile` issue to be addressed or ignored as I mentioned above.
Whether that warning will fail the build or not is still unclear to me.

westonruter commented:
Clearly my `dev-develop` didn't work. My Composer expertise fails me.

jrfnl commented:
> Clearly my dev-develop didn't work. My Composer expertise fails me.

Nah, it was just Friday afternoon ;-)
You've updated the Composer config, but not run `composer update` again, so the `composer.lock` file which you committed is out of date.
Either the `composer.lock` file should not be committed or else, it should be inline with the config. :smiley_cat:

jrfnl commented:
On another note, if we're using Composer now anyway, shouldn't `dev-lib` be pulled in as a Composer requirement as well and removed as a submodule ?

westonruter commented:
Yeah, installing dev-lib via Composer would be a good idea too. Or via NPM: https://github.com/xwp/ampconf/blob/2f4279f59b2c5942bfebbdc76e66ae217cc44e92/package.json#L23

jrfnl commented:
Oh.. and thinking about it some more... what about adding a `.gitattributes` file to make sure all the dev files don't get packaged up in the releases ?

westonruter committed:
Build/PHPCS: exclude Generic.Files.OneClassPerFile.MultipleFound

westonruter commented:
@jrfnl Sounds good. Normally I've been using [grunt-wp-deploy](https://www.npmjs.com/package/grunt-wp-deploy) lately, but that's somewhat overkill for this plugin.

jrfnl commented:
@westonruter I'll add it. `.gitattributes` works independently of how you deploy to the plugin repo, so it will also benefit people who download a release here on GH or would pull the plugin in via Composer.

Would it be ok if I cleaned up the branch a little ? (squash a couple of commits and such)

westonruter commented:
@jrfnl Feel free to amend the branch any way you wish!

jrfnl committed:
Composer: update the minimum versions of CS/QA dependencies

... and switch to using PHPCompatibilityWP.

jrfnl committed:
Composer: declare `wp-dev-lib` as a dependency

... and remove the `wp-dev-lib` submodule.

jrfnl committed:
Add .gitattributes file

jrfnl commented:
So... I finally got round to adjusting this PR. Asides from what was discussed above, some additional updates were in order due to new releases of dev tools in use.

Changes I've made compared to where we left off in January:
* Reorganized the branch so the commits are atomic.
* Added a `.gitattributes` file.
* Removed `PHPCS_IGNORE` from the `.dev-lib` config as the directives in the custom ruleset should be used instead.
* Removed the `xwp-dev-lib` submodule in favour of declaring it as a composer dependency.
* Added better caching of Composer downloads to the Travis script.
* Removed references to node from the Travis script as it not used in this repo.
* Switched the PHPCS run from PHP 7.0 to PHP 7.2 in the Travis script.
* Composer: Use WPCS 1.0.0
* Composer + PHPCS config: use `PHPCompatibilityWP` instead of `PHPCompatibility`.
* PHPCS ruleset:
    - Improved the inline documentation.
    - Added `basepath` directive.
    - Use new PHPCS 3.3.0+ property array format.
    - Removed the `<exclude-pattern>`s as the ruleset explicitly only includes the one file, so they has no function.

There is one curiosity causing the Travis PHP 5.2 build to fail, which I believe is a bug in the `wp-dev-lib` library: when on PHP 5.2, the `wp-dev-lib` `install_tools` command tries to `composer install`.

This shouldn't happen as:
1. `composer` is not in the `DEV_LIB_ONLY` environment variable which is applicable to PHP 5.2.
2. AFAICS, `wp-dev-lib` checks the PHP version and sets `composer` to `skip` when on PHP 5.2.

So, it is quite unclear to me why the script still tries to `composer install`. @westonruter Have you got any bright ideas on how to solve this ?

That issue aside, this PR is ready for merge.

westonruter commented:
> There is one curiosity causing the Travis PHP 5.2 build to fail

A good workaround was devised by @kienstra in https://github.com/xwp/pwa-wp/commit/082954047e1702c34dba75d05665df27d234156e

jrfnl commented:
@westonruter iMHO the bug should be fixed in the `dev-lib` library rather than get people to use esoteric work-arounds for a situation which shouldn't ever occur in the first place.
